compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -456,6 +458,7 @@ public void visit(BLangFunction funcNode) {
         }
 
         validateParams(funcNode);
+        logErrorsForInferredArrays(funcNode.returnTypeNode);
```

Comment:
I believe there's room for improvement here.

What I thought was that the code analyzer would call `analyzeTypeNode` for all the type nodes wherever applicable. That should have been done already (not related to this change). For example, we would have `analyzeTypeNode(funcNode.returnTypeNode, env);` in this visit method. And similarly visits for param types.

And in the visit for `BLangArrayType` we do the entire validation looking at the parent and identifying the context. That way we wouldn't have to add special checks in arbitrary places. I believe we should be able to get rid of the `isValidInferredArray` field also.

Reply:
+1. But I have noticed that we need a separate check to handle the return type of a function, since the visitor of `BLangArray` is not called when there are array types in return type node.

Reply:
Will update as we discussed.